### ISSUE TITLE
Segfault upon @everywhere using TensorFlow

### ISSUE THREAD

**Issue by baggepinnen** [malmaud/TensorFlow.jl#137]
I receive segmentation faults whenever I load TF.jl on more than one worker. Maybe this is a hard or even impossible problem to solve, but it would be nice to have the option of running many experiments etc. in parallel. If I on the other hand launch two separate instances of julia, tensorflow runs smoothly on both simultaneously.
`julia -p 1`
```
julia> @everywhere using TensorFlow
WARNING: replacing module TensorFlow.

signal (11): Segmentation fault
while loading no file, in expression starting on line 0
unknown function (ip: 0x7f7d6779b265)
unknown function (ip: 0x7f7d6779ba8c)
unknown function (ip: 0x7f7d6779b74d)
unknown function (ip: 0x7f7d6779ba8c)
unknown function (ip: 0x7f7d6779b74d)
unknown function (ip: 0x7f7d6779b74d)
unknown function (ip: 0x7f7d6779b74d)
unknown function (ip: 0x7f7d6779b74d)
unknown function (ip: 0x7f7d6779b74d)
unknown function (ip: 0x7f7d6779cf8d)
unknown function (ip: 0x7f7d6779c166)
unknown function (ip: 0x7f7d6779b74d)
unknown function (ip: 0x7f7d6779b74d)
unknown function (ip: 0x7f7d6779bc81)
unknown function (ip: 0x7f7d6779b74d)
unknown function (ip: 0x7f7d6779e1bb)
jl_gc_collect at /usr/bin/../lib64/libjulia.so.0.5 (unknown line)
jl_gc_pool_alloc at /usr/bin/../lib64/libjulia.so.0.5 (unknown line)
jl_gc_alloc at /usr/bin/../lib64/libjulia.so.0.5 (unknown line)
jl_alloc_svec_uninit at /usr/bin/../lib64/libjulia.so.0.5 (unknown line)
jl_f_svec at /usr/bin/../lib64/libjulia.so.0.5 (unknown line)
jl_apply_generic at /usr/bin/../lib64/libjulia.so.0.5 (unknown line)
unknown function (ip: 0x7f7d6776b2fd)
unknown function (ip: 0x7f7d6776a3ed)
jl_interpret_toplevel_expr_in at /usr/bin/../lib64/libjulia.so.0.5 (unknown line)
unknown function (ip: 0x7f7d677c7b43)
unknown function (ip: 0x7f7d677d1125)
unknown function (ip: 0x7f7d677d5a38)
unknown function (ip: 0x7f7d677d7c4e)
unknown function (ip: 0x7f7d677d88fa)
unknown function (ip: 0x7f7d677d9973)
unknown function (ip: 0x7f7d677e4a38)
unknown function (ip: 0x7f7d677e6f6b)
unknown function (ip: 0x7f7d677e9a22)
unknown function (ip: 0x7f7d67756d17)
jl_apply_generic at /usr/bin/../lib64/libjulia.so.0.5 (unknown line)
unknown function (ip: 0x7f7d67784949)
unknown function (ip: 0x7f7d6777bc9c)
jl_restore_incremental_from_buf at /usr/bin/../lib64/libjulia.so.0.5 (unknown line)
#762 at ./multi.jl:1909
jl_apply_generic at /usr/bin/../lib64/libjulia.so.0.5 (unknown line)
jl_f__apply at /usr/bin/../lib64/libjulia.so.0.5 (unknown line)
#622 at ./multi.jl:1417
run_work_thunk at ./multi.jl:1001
run_work_thunk at ./multi.jl:1010 [inlined]
#597 at ./event.jl:68
unknown function (ip: 0x7f7b3cdeec1f)
jl_apply_generic at /usr/bin/../lib64/libjulia.so.0.5 (unknown line)
unknown function (ip: 0x7f7d677742fe)
unknown function (ip: 0xffffffffffffffff)
Allocations: 7338783 (Pool: 7337520; Big: 1263); GC: 9
Worker 2 terminated.ERROR: ProcessExitedException()
 in (::##15#17)() at ./multi.jl:1957
 in macro expansion; at ./multi.jl:1968 [inlined]
 in anonymous at ./<missing>:?
 in eval_user_input(::Any, ::Base.REPL.REPLBackend) at ./REPL.jl:64
 in macro expansion at ./REPL.jl:95 [inlined]
 in (::Base.REPL.##3#4{Base.REPL.REPLBackend})() at ./event.jl:68


ERROR (unhandled task failure): EOFError: read end of file
```

**Comment by malmaud** [comment]
The issue is that TensorFlow.jl spawns a new Julia process upon initialization to communicate with Python and the logic for that hasn't been coding to deal with the case of the user also spawning new Julia processes. 

It should be solvable in principle, and I should get to it sometime (and eventually, I won't need to create the Python process at all). In the meanwhile, you can get  some TensorFlow parallelism from a single Julia process by using the QueueRunners.

**Comment by malmaud** [comment]
I've made progress in https://github.com/malmaud/TensorFlow.jl/commit/9f32de3cfb7c5c3ac24784a6ccfc11ef6861e20d.

Now you can safely load TensorFlow on workers, but only the main process can compute gradients. I'll see about lifting that restriction.

Also, I think `@everywhere using` is not the syntax you want. `using` already loads the package on all processes, so `@everywhere using` will actually load the package `nprocs()` time on each worker.

**Comment by baggepinnen** [comment]
I also noticed the issue with `@everywhere using`, however, `using` alone only imports on workers, the names are not brought into scope. I solved it by `@everywhere include("..")` where the using statements are run in the included file, but never on the main process. This works well for some of my other problems, hopefully on TensorFlow also in the future. Thanks for your efforts!

**Comment by malmaud** [comment]
OK, I think I basically fixed this. There still might be some possible race conditions I didn't consider.

**Comment by malmaud** [comment]
Great!
